src/librustc_typeck/check/mod.rs

```diff
@@ -1522,6 +1517,14 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         self.tables.borrow_mut().node_types.insert(node_id, ty);
     }
 
+    #[inline]
+    pub fn write_ty_expr(&self, node_id: ast::NodeId, ty: Ty<'tcx>) {
```

Comment:
@eddyb That doesn't seem to work. I changed `write_ty` to this:

```rust
#[inline]
pub fn write_ty(&self, node_id: ast::NodeId, ty: Ty<'tcx>) {
    debug!("write_ty({}, {:?}) in fcx {}",
           node_id, ty, self.tag());
    self.tables.borrow_mut().node_types.insert(node_id, ty);

    // Add adjustments to !-expressions
    if ty.is_never() {
        if let hir::map::NodeExpr(_) = self.tcx.map.find(node_id).unwrap() {
            self.write_adjustment(node_id, adjustment::AdjustNeverToAny(self.next_diverging_ty_var()));
        }
    }
}
```

And I'm getting "called `Option::unwrap()` on a `None` value". If `map.find` is returning `None` then I don't know whether to write the adjustment or not.